Checksum Cleanup

schisamo opened the pull request:
This pull request adds the ability to delete a list of checksums on S3/Bookshelf.  The deletion is done in chunked groups with parallel deletion of the checksums within each group.
- **chef_s3:delete_checksums/2** - Given a list of checksums, chunk it into (configurable) groups and
  delete the checksums in each chunk in parallel.
- **s3_ops:delete_checksums/2** - Entry point used by codebase as a whole into S3 
  checksum deletion logic.
- additional test coverage for new logic

schisamo committed:
rename s3_metadata module to more general s3_ops

This is in prep of moving some checksum delete operations into s3_ops.

schisamo committed:
add erlware_commons as a dep

schisamo committed:
add s3_ops:delete/2 function

Given a list of checksums, chunk it into (configurable) groups and
delete the checksums in each chunk in parallel.

schisamo committed:
add chef_s3:delete_checksums/2

Entry point used by codebase as a whole into S3 
checksum deletion logic.

schisamo committed:
additional test coverage for checksum deletion logic

schisamo commented:
^^ @seth @christophermaier feedback when you have a sec.

seth commented on src/s3_ops.erl:
```diff
@@ -41,11 +43,68 @@
+    %% parallelizing ALL THE THINGS!!
+    Chunks = chunk_list(Checksums, chunk_size()),
+    %% Delete the checksums in each chunk in parrallel
+    Results = lists:map(fun(Chunk) ->
```
I think this will read better as a list comprehension, but see review patch for a possible different approach as well.

seth commented on src/s3_ops.erl:
```diff
@@ -41,11 +43,68 @@
+    FlattenedResults = lists:flatten(Results),
+    %% Process the flattened results, splitting them into success checksums,
+    %% missing checksums, timeout count and error count
+    #state{ok=Ok, missing=Missing, timeouts=Timeouts, errors=Errors} =
```
I think the state record isn't adding much. See what you think of the version in the attached patch.

seth commented:
Overall, this is looking good.

There are a few places where the indentation is wrong and a few places where it is "wrong" (i.e, isn't what Emacs would do).

In terms of the module name change, I'm not sure how we ended up with the non "chef_" prefixed modules. I think we should change it to chef_s3_ops. I sort of liked Eric's convention of only naming external API modules with the app name prefix and creating a short prefix for internal modules (e.g. chf_). Obviously something we would move to after some discussion and not in this PR. (we should make a note to rename deep_merge).

I think the lists:foldl in delete/2 is easier to read without the state record and I took a stab showing that along with swapping chunk_list with chunk_map since that's actually what is happening here. Anyhow, have a look and see if you like any of it:

https://gist.github.com/580b5921156842e101d4

Finally, for the unit tests, you can disable tty logging so the error logger message won't appear. There are examples of that elsewhere and I think that's worth doing to have quieter test runs.

schisamo commented:
@seth Thanks for the feedback!  Definetly like the suggestions (especially chunk_map/3).

Regarding the bad indention, we should probably document style conventions like this on the wiki vs relying on what may or may not default behavior of one editor. This will become especially important as this code has become open-source and our potential developer community becomes larger.

schisamo committed:
add chunk_map/4 function; chunks list and applies function in one go.

schisamo committed:
simplify result processing, nuke the state record

schisamo committed:
Properly indent ALL THE THINGS!

schisamo committed:
Disable TTY logging output during chef_s3 tests.

seth commented on test/chef_s3_tests.erl:
```diff
@@ -147,16 +153,48 @@ check_test_() ->
      end,
      fun(_OrgId) ->
+             %% Re-enable logging
+             error_logger:tty(false),
```
We want `true` here.

seth commented:
Looking good. Thoughts on renaming `s3_ops` to `chef_s3_ops` ?

schisamo commented:
I'm all for the rename..but you said: _Obviously something we would move to after some discussion and not in this PR._ above :)

seth commented:
Ah, well, fair enough :) I was referring more to the change to using a prefix for internal only modules, like chf_. I don't see any reason not to use chef_s3_ops for now. But fine with leaving it to later as well.

In any case, for this PR: please to merge! (haven't had a chance to look at chef_db yet)

schisamo commented:
I'll do the rename and fix that `error_logger:tty` slip before merging.

seth commented:
rock on.

schisamo committed:
rename s3_ops -> chef_s3_ops

idiomatically we should try and make module name prefix match the 
project prefix.

schisamo committed:
rename deep_merge -> chef_deep_merge
    
idiomatically we should try and make module name prefix match the
project prefix.

schisamo committed:
remove noisy logger info statement

seth commented:
Not sure why this is still open. This branch has been merged and deleted as far as I can tell.
